modules/sitemaps/sitemaps.php

```diff
@@ -1,792 +1,1865 @@
+			// Catch news sitemap xsl.
+			if ( preg_match( $regex['news-style'], $request_uri ) ) {
+				$this->serve_raw_and_die(
+					'text/xml',
+					Jetpack_Sitemap_Stylist::news_sitemap_xsl()
+				);
+			}
+		}
+
+		// URL did not match any sitemap patterns.
+		return;
 	}
 
-	// Compatibility with PHP 5.3 and older
-	if ( ! defined( 'ENT_XML1' ) ) {
-		define( 'ENT_XML1', 16 );
+	/**
+	 * The complete uri at which sitemaps are served.
+	 */
+	private function the_jetpack_sitemap_uri() {
+		return home_url(
+			$this->read_sitemap_location_option()
+		);
+	}
+
+	/**
+	 * The complete path at which sitemaps are served.
+	 */
+	private function the_jetpack_sitemap_path() {
+		$parsed_url = wp_parse_url( $this->the_jetpack_sitemap_uri() );
+
+		if ( isset( $parsed_url['path'] ) ) {
+			$path = $parsed_url['path'];
+		} else {
+			$path = '';
+		}
+
+		return $path;
 	}
 
 	/**
-	 * Filter the post types that will be included in sitemap.
+	 * Callback for adding sitemap-interval to the list of schedules.
 	 *
-	 * @module sitemaps
+	 * @since 4.5.0
 	 *
-	 * @since 3.9.0
+	 * @param array $schedules The array of WP_Cron schedules.
+	 */
+	public function callback_add_sitemap_schedule( $schedules ) {
+		$schedules['sitemap-interval'] = array(
+			'interval' => Jetpack_Sitemap_Manager::SITEMAP_INTERVAL,
+			'display'  => __( 'Sitemap Interval', 'jetpack' ),
+		);
+		return $schedules;
+	}
+
+	/**
+	 * Add actions to schedule sitemap generation.
+	 * Should only be called once, in the constructor.
 	 *
-	 * @param array $post_types Array of post types.
+	 * @access private
+	 * @since 4.5.0
 	 */
-	$post_types    = apply_filters( 'jetpack_sitemap_post_types', array( 'post', 'page' ) );
+	private function schedule_sitemap_generation() {
+		// Add cron schedule.
+		add_filter( 'cron_schedules', array( $this, 'callback_add_sitemap_schedule' ) );
+
+		add_action(
+			'jp_sitemap_cron_hook',
+			array( $this, 'build_all_sitemaps' )
+		);
+
+		if ( ! wp_next_scheduled( 'jp_sitemap_cron_hook' ) ) {
+			wp_schedule_event(
+				time(),
+				'sitemap-interval',
+				'jp_sitemap_cron_hook'
+			);
+		}
 
-	$post_types_in = array();
-	foreach ( (array) $post_types as $post_type ) {
-		$post_types_in[] = $wpdb->prepare( '%s', $post_type );
+		return;
 	}
-	$post_types_in = join( ",", $post_types_in );
 
-	// use direct query instead because get_posts was acting too heavy for our needs
-	//$posts = get_posts( array( 'numberposts'=>1000, 'post_type'=>$post_types, 'post_status'=>'published' ) );
-	$posts = $wpdb->get_results( "SELECT ID, post_type, post_modified_gmt, comment_count FROM $wpdb->posts WHERE post_status='publish' AND post_type IN ({$post_types_in}) ORDER BY post_modified_gmt DESC LIMIT 1000" );
-	if ( empty( $posts ) ) {
-		status_header( 404 );
+	/**
+	 * Callback to add sitemap to robots.txt.
+	 *
+	 * @access public
+	 * @since 4.5.0
+	 */
+	public function callback_action_do_robotstxt() {
+		$url = $this->the_jetpack_sitemap_uri() . '/sitemap.xml';
```

Comment:
The latest commit handles plain and PATHINFO permalinks. Plain permalinks are served at `/?jetpack-sitemap=sitemap.xml`.